File: public/java/020-next-permutation.java
/*
 * LeetCode Problem #31: Next Permutation
 * Difficulty: Medium
 * Link: https://leetcode.com/problems/next-permutation/
 */
import java.util.*;

class Solution {

    // APPROACH 1: STL next_permutation (Library) | O(n) time | O(1) space
    // EXPLAIN: Replicate what C++ std::next_permutation does — for completeness; Java has no direct equivalent.
    // WHEN: Reference implementation to understand the algorithm before coding it manually.
    public void nextPermutation_STL(int[] nums) {
        // Java has no built-in next_permutation; implement it exactly as approach 2
        nextPermutation(nums);
    }

    // APPROACH 2: Classic In-Place | O(n) time | O(1) space
    // EXPLAIN: Find the rightmost descent, swap with the next larger element to its right, then reverse the suffix.
    // WHEN: The only practical approach — exploits the structure of permutation ordering directly.
    public void nextPermutation(int[] nums) {
        int n = nums.length;

        // Step 1: Find rightmost index i where nums[i] < nums[i+1]
        int i = n - 2;
        while (i >= 0 && nums[i] >= nums[i + 1]) {
            i--;
        }

        if (i >= 0) {
            // Step 2: Find rightmost index j where nums[j] > nums[i]
            int j = n - 1;
            while (nums[j] <= nums[i]) {
                j--;
            }
            // Step 3: Swap nums[i] and nums[j]
            int temp = nums[i];
            nums[i] = nums[j];
            nums[j] = temp;
        }

        // Step 4: Reverse the suffix starting at i+1
        int left = i + 1, right = n - 1;
        while (left < right) {
            int temp = nums[left];
            nums[left] = nums[right];
            nums[right] = temp;
            left++;
            right--;
        }
    }

    // APPROACH 3: Explicit Reverse (No Arrays.sort) | O(n) time | O(1) space
    // EXPLAIN: Same four-step algorithm as approach 2 but the suffix reversal is spelled out without helpers.
    // WHEN: When you want to be explicit about each operation and avoid any library call.
    public void nextPermutation_ExplicitReverse(int[] nums) {
        int n = nums.length;
        int i = n - 2;

        while (i >= 0 && nums[i] >= nums[i + 1]) i--;

        if (i >= 0) {
            int j = n - 1;
            while (nums[j] <= nums[i]) j--;
            int tmp = nums[i]; nums[i] = nums[j]; nums[j] = tmp;
        }

        // Manual reverse of suffix
        int lo = i + 1, hi = n - 1;
        while (lo < hi) {
            int tmp = nums[lo]; nums[lo] = nums[hi]; nums[hi] = tmp;
            lo++; hi--;
        }
    }

    // APPROACH 4: Two-Pass Scan (Verbose / Educational) | O(n) time | O(1) space
    // EXPLAIN: Same as approach 2 but uses descriptive variable names (pivot, swapIdx) to highlight the four conceptual steps.
    // WHEN: Teaching or explaining the algorithm — very readable for code review or whiteboard discussions.
    public void nextPermutation_TwoPass(int[] nums) {
        int n = nums.length;

        // Pass 1: find "dip" — last index where ascending order breaks
        int pivot = -1;
        for (int i = n - 2; i >= 0; i--) {
            if (nums[i] < nums[i + 1]) { pivot = i; break; }
        }

        if (pivot == -1) {
            // Entirely non-increasing → return smallest permutation
            Arrays.sort(nums);
            return;
        }

        // Pass 2: find smallest in suffix that is greater than pivot element
        int swapIdx = -1;
        for (int j = n - 1; j > pivot; j--) {
            if (nums[j] > nums[pivot]) { swapIdx = j; break; }
        }

        int tmp = nums[pivot]; nums[pivot] = nums[swapIdx]; nums[swapIdx] = tmp;

        // Reverse suffix
        int lo = pivot + 1, hi = n - 1;
        while (lo < hi) {
            tmp = nums[lo]; nums[lo] = nums[hi]; nums[hi] = tmp;
            lo++; hi--;
        }
    }

    // APPROACH 5: Sort-Based | O(n log n) time | O(1) space
    // EXPLAIN: Same pivot-and-swap logic but sort the suffix instead of reversing it — correct because suffix is descending.
    // WHEN: Slightly less optimal than reversing but works; shows understanding that sorting a desc array = O(n).
    public void nextPermutation_SortBased(int[] nums) {
        int n = nums.length;
        int i = n - 2;

        while (i >= 0 && nums[i] >= nums[i + 1]) i--;

        if (i >= 0) {
            int j = n - 1;
            while (nums[j] <= nums[i]) j--;
            int tmp = nums[i]; nums[i] = nums[j]; nums[j] = tmp;
        }

        // Sort suffix (works because it is in descending order after the swap)
        Arrays.sort(nums, i + 1, n);
    }
}

// Made with Bob

File: public/python/020-next-permutation.py
# LeetCode Problem #31: Next Permutation
# Difficulty: Medium
# Link: https://leetcode.com/problems/next-permutation/

# ─────────────────────────────────────────────
# APPROACH 1: STL next_permutation (Library) | O(n) time | O(1) space
# EXPLAIN: Use Python's itertools equivalent — find the next permutation in lexicographic order.
# WHEN: Reference / quick-check; not available as a single built-in so we replicate the algorithm.

def next_permutation_stl(nums: list[int]) -> None:
    """Modifies nums in-place (delegates to classic in-place algorithm)."""
    # Python has no direct next_permutation — replicate the classic algorithm
    n = len(nums)
    i = n - 2
    while i >= 0 and nums[i] >= nums[i + 1]:
        i -= 1
    if i >= 0:
        j = n - 1
        while nums[j] <= nums[i]:
            j -= 1
        nums[i], nums[j] = nums[j], nums[i]
    lo, hi = i + 1, n - 1
    while lo < hi:
        nums[lo], nums[hi] = nums[hi], nums[lo]
        lo += 1; hi -= 1


# ─────────────────────────────────────────────
# APPROACH 2: Classic In-Place | O(n) time | O(1) space
# EXPLAIN: Find the rightmost descent, swap with the next-larger element to its right, then reverse the suffix.
# WHEN: The canonical O(n) in-place algorithm — always use this; it's the expected solution.

def next_permutation_inplace(nums: list[int]) -> None:
    """Modifies nums in-place."""
    n = len(nums)
    # Step 1: find the largest index i such that nums[i] < nums[i+1]
    i = n - 2
    while i >= 0 and nums[i] >= nums[i + 1]:
        i -= 1

    if i >= 0:
        # Step 2: find the largest index j > i such that nums[i] < nums[j]
        j = n - 1
        while nums[j] <= nums[i]:
            j -= 1
        # Step 3: swap
        nums[i], nums[j] = nums[j], nums[i]

    # Step 4: reverse the suffix starting at i+1
    lo, hi = i + 1, n - 1
    while lo < hi:
        nums[lo], nums[hi] = nums[hi], nums[lo]
        lo += 1; hi -= 1


# ─────────────────────────────────────────────
# APPROACH 3: Explicit Reverse | O(n) time | O(1) space
# EXPLAIN: Identical to approach 2 but spells out the suffix reversal explicitly without any helper call.
# WHEN: When you want every step in plain view — useful on a whiteboard.

def next_permutation_explicit(nums: list[int]) -> None:
    """Modifies nums in-place."""
    n = len(nums)
    i = n - 2
    while i >= 0 and nums[i] >= nums[i + 1]:
        i -= 1
    if i >= 0:
        j = n - 1
        while nums[j] <= nums[i]:
            j -= 1
        nums[i], nums[j] = nums[j], nums[i]
    # Manual reverse
    lo, hi = i + 1, n - 1
    while lo < hi:
        nums[lo], nums[hi] = nums[hi], nums[lo]
        lo += 1
        hi -= 1


# ─────────────────────────────────────────────
# APPROACH 4: Two-Pass Scan (Educational / Verbose) | O(n) time | O(1) space
# EXPLAIN: Same four-step logic with descriptive names (pivot, swap_idx) to highlight each conceptual step.
# WHEN: Teaching or code review; very readable for explaining the algorithm.

def next_permutation_two_pass(nums: list[int]) -> None:
    """Modifies nums in-place."""
    n = len(nums)
    pivot = -1
    for i in range(n - 2, -1, -1):
        if nums[i] < nums[i + 1]:
            pivot = i
            break

    if pivot == -1:
        nums.sort()
        return

    swap_idx = -1
    for j in range(n - 1, pivot, -1):
        if nums[j] > nums[pivot]:
            swap_idx = j
            break

    nums[pivot], nums[swap_idx] = nums[swap_idx], nums[pivot]

    lo, hi = pivot + 1, n - 1
    while lo < hi:
        nums[lo], nums[hi] = nums[hi], nums[lo]
        lo += 1; hi -= 1


# ─────────────────────────────────────────────
# APPROACH 5: Sort-Based | O(n log n) time | O(1) space
# EXPLAIN: Same pivot-and-swap; sort the suffix instead of reversing — correct because suffix is already descending.
# WHEN: Slightly slower (O(n log n) vs O(n)) but shows that sorting descending = reversing.

def next_permutation_sort(nums: list[int]) -> None:
    """Modifies nums in-place."""
    n = len(nums)
    i = n - 2
    while i >= 0 and nums[i] >= nums[i + 1]:
        i -= 1
    if i >= 0:
        j = n - 1
        while nums[j] <= nums[i]:
            j -= 1
        nums[i], nums[j] = nums[j], nums[i]
    # Sort suffix (equivalent to reversing because it is already descending)
    nums[i + 1:] = sorted(nums[i + 1:])


# ─────────────────────────────────────────────
if __name__ == "__main__":
    import copy

    cases = [
        ([1, 2, 3], [1, 3, 2]),
        ([3, 2, 1], [1, 2, 3]),
        ([1, 1, 5], [1, 5, 1]),
        ([1],       [1]),
    ]
    fns = [
        next_permutation_stl,
        next_permutation_inplace,
        next_permutation_explicit,
        next_permutation_two_pass,
        next_permutation_sort,
    ]
    for nums, expected in cases:
        for fn in fns:
            m = copy.deepcopy(nums)
            fn(m)
            assert m == expected, f"{fn.__name__} failed on {nums}: got {m}"

    print("All tests passed.")

# Made with Bob
